config-model/src/main/java/com/yahoo/vespa/model/container/component/BindingPattern.java

```diff
@@ -0,0 +1,127 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.model.container.component;
+
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * URI binding pattern used by filter and handler bindings.
+ *
+ * There are two types of binding; user generated and model generated bindings.
+ * - User generated bindings are bindings which are constructed from directly from 'binding' elements from services.xml
+ * - Model generated bindings are binding which are implicitly constructed by the model, e.g built-in handlers.
+ *
+ * @author bjorncs
+ */
+public class BindingPattern implements Comparable<BindingPattern> {
+
+    private static final Pattern BINDING_PATTERN =
+            Pattern.compile("([^:]+)://([^:/]+)(:((\\*)|([0-9]+)))?(/.*)", Pattern.UNICODE_CASE | Pattern.CANON_EQ);
+
+    private final String scheme;
+    private final String host;
+    private final String port;
+    private final String path;
+    private final boolean isUserGenerated;
```

Comment:
The boolean variable and the verbose static factory method names are indeed suboptimal. I like your proposal with the "user" vs "system" distinction (instead of the confusing "model" prefix). I prefer the class hierarchy solution given that the "type"/"source" aspect is a fundamental for handling of bindings in the model (which I think it is).